### PR TITLE
feat(model schema): Adding samples of the Model schema for energy models

### PR DESCRIPTION
**Pull request by chriswmackey**
@tanushree04 , @macumber , and @mostaphaRoudsari ,

This PR includes 3 samples of the Model schema (2 for single zone energy models and one that is multizone) so that @tanushree04 has a good reference point for the work that will be done over at least the next 3 weeks.

As you can infer, the minimum amount of Python code for generating the model schema files is done and so I can generate a wide variety of cases with ease.  The one thing that is missing right now from the examples and the Python code is that there are no Loads, Schedules, or ProgramTypes.  However, the first sample (1_single_zone_tiny_house.json) includes all of the 5 geometry objects.  Also, the samples use all of the construction-related schemas we developed over the last few months (ConstructionSetAbridged, OpaqueConstructionAbridged, WindowConstructionAbridged, and several materials).  So these models should be simulate-able, albeit in a fully passive manner.

Now that we have these samples, I will begin work on some diagrams that will make the structure of the geometry objects clearer (Rooms, Faces, Shades, Apertures, Doors) .

Lastly, I'll note that, pending some review from @mostaphaRoudsari , it is possible that we make some tweaks to the schema in the coming weeks.  However, the general structure you see in these files has emerged from a lot of discussions and debates over the last few months, meaning we have good reasons for most of the decisions we made and so this general structure is far less likely to change.

### CODE REVIEW

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 1]
```diff
@@ -0,0 +1,961 @@
+{
+    "type": "Model",
+    "name": "TinyHouse",
+    "name_original": "Tiny House",
+    "properties": {
+        "type": "ModelProperties",
+        "energy": {
+            "type": "ModelEnergyProperties",
+            "global_construction_set": {
```
Comment (macumber) [inline tests/fixtures/1_single_zone_tiny_house.json]: Could be possible to make this a reference into the construction_sets array

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 2]
```diff
@@ -0,0 +1,961 @@
...
+                        "Generic Window Air Gap",
+                        "Generic Clear Glass",
+                        "Generic Window Air Gap",
+                        "Generic Clear Glass"
+                    ]
+                }
+            ],
+            "materials": [
+                {
+                    "type": "EnergyMaterial",
```
Comment (macumber) [inline tests/fixtures/1_single_zone_tiny_house.json]: There is no abridged version of an EnergyMaterial?  Abridged means "references external objects" and non-abdriged means "all properties nested"?

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 3]
```diff
@@ -0,0 +1,961 @@
...
+                    "visible_absorptance": 0.8
+                }
+            ]
+        }
+    },
+    "rooms": [
+        {
+            "type": "RoomAbridged",
+            "name": "TinyHouseZone",
+            "name_original": "Tiny House Zone",
```
Comment (macumber) [inline tests/fixtures/1_single_zone_tiny_house.json]: is name_original the name in grasshopper or a display name?  I'd suggest using the key `display_name` instead of `name_original`.

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 4]
```diff
@@ -0,0 +1,961 @@
...
+                    "solar_absorptance": 0.75,
+                    "visible_absorptance": 0.8
+                }
+            ]
+        }
+    },
+    "rooms": [
+        {
+            "type": "RoomAbridged",
+            "name": "TinyHouseZone",
```
Comment (macumber) [inline tests/fixtures/1_single_zone_tiny_house.json]: This is the name you want used in E+?  `name` will conform to all the uniqueness and character set requirements of E+?

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 5]
```diff
@@ -0,0 +1,961 @@
...
+                    ],
+                    [
+                        5.0,
+                        0.0,
+                        0.0
+                    ]
+                ]
+            },
+            "properties": {
+                "type": "FaceProperties",
```
Comment (macumber) [inline tests/fixtures/1_single_zone_tiny_house.json]: Why are some properties (e.g. `energy::construction` in here but others are not `boundary_condition`)?

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 6]
```diff
@@ -0,0 +1,961 @@
...
+            "faces": [
+                "TinyHouseZone_Bottom",
+                "TinyHouseZone_Front",
+                "TinyHouseZone_Right",
+                "TinyHouseZone_Back",
+                "TinyHouseZone_Left",
+                "TinyHouseZone_Top"
+            ],
+            "indoor_shades": [
+                "TinyHouseZone_Back_InShd0"
```
Comment (macumber) [inline tests/fixtures/1_single_zone_tiny_house.json]: Are indoor shades things like cubicle walls?  These would be InteriorPartitionSurfaces in OS.

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 7]
```diff
@@ -0,0 +1,961 @@
...
+            "face_type": {
+                "type": "Floor"
+            },
+            "boundary_condition": {
+                "type": "Ground",
+                "sun_exposure": "NoSun",
+                "wind_exposure": "NoWind",
+                "view_factor": "autocalculate"
+            },
+            "parent": "TinyHouseZone"
```
Comment (macumber) [inline tests/fixtures/1_single_zone_tiny_house.json]: Suggest having the child reference the parent or the parent reference the children.  Having two way references can lead to data being out of sync.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 1 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: Now that you mention it, I think you are right that a reference is the right way to do it.  I think I'll go ahead and implement it once @mostaphaRoudsari gets the chance to do an initial review.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 7 above)
Comment (macumber) [reply tests/fixtures/1_single_zone_tiny_house.json]: Note, you can have the two way link in memory when you are working with live Python objects.  I just wouldn't write both out to disk.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 2 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: Yes, "Abridged" means that not all of the information needed to completely re-serialize the object to its Python version is contained within the dictionary.  In other words, there are references to other objects by name and not as nested children objects.  Because Materials always have all of the information to re-create them, there's no need for an Abridged schema for them.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 4 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: Yes, you can see that this is clarified in the Google slides presentation of the diagrams that I sent to the slack channel.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 3 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: That is a good suggestion and let's get @mostaphaRoudsari 's thoughts on it.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 6 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: Essentially, yes. In this sample here they refer to the interior half of a light shelf but they are also intended to describe all types of interior furniture (including cubicle walls).  They will have more significance to Radiance simulations and detailed thermal comfort mapping than they do to the exported EnergyPlus model but, if you have a way to include them in the OpenStudio model, we should consider using it.  One thing we are sure we would like is to have the Room's `outdoor_shades` assigned as `Zone:Shading:Detailed` rather than `Site:Shading:Detailed` in the ultimate E+ model.  We can discuss some of the ways to make this happen on the call Wednesday.

Comment (macumber) [reply tests/fixtures/1_single_zone_tiny_house.json]: Cool, it would be good for us if we can differentiate between light shelves and interior partitions.  Is it possible to associate light shelves with a particular window?

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 5 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: `boundary_condition` is something that we realized needs to be shared between Radiance and Energy models since, particularly when it comes to 3-phase daylight, you need to know when an Aperture is next to another room or directly sees the sky.  So the `face_type` and `boundary_condition` exist on the core Face object and only the properties specifically related to energy go here under this section.  Right now, this is only the construction.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 7 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: @mostaphaRoudsari  and I have had long conversations about this, which have also included some input from Theo , who provided feedback in terms of what he thought was easiest for editing the schema.  Let's plan to have a longer discussion about this on Wednesday.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 6 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: At the moment, the schema doesn't support shades for individual Apertures but @mostaphaRoudsari thinks that we will need this for Radiance and he [has opened an issue about it](https://github.com/ladybug-tools/honeybee-core/issues/11). We can discuss Wednesday but maybe for now we just focus on implementing Room-assigned shade and we can add Aperture-assigned shade to the schema afterwards.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 4 above)
Comment (tanushree04) [reply tests/fixtures/1_single_zone_tiny_house.json]: @chriswmackey is there a character limit (le < 100)  for the `name` as well?

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 8]
```diff
@@ -0,0 +1,961 @@
...
+                        0.5,
+                        2.9999999
+                    ]
+                ]
+            },
+            "properties": {
+                "type": "ShadeProperties",
+                "energy": {
+                    "type": "ShadeEnergyPropertiesAbridged",
+                    "diffuse_reflectance": 0.2,
```
Comment (tanushree04) [inline tests/fixtures/1_single_zone_tiny_house.json]: @chriswmackey are there any default values for these properties?

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 9]
```diff
@@ -0,0 +1,961 @@
...
+                    "specular_reflectance": 0.0,
+                    "transmittance": 0.0
+                }
```
Comment (tanushree04) [inline tests/fixtures/1_single_zone_tiny_house.json]: @chriswmackey should the `transmittance_schedule`  name lack special characters?

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 10]
```diff
@@ -0,0 +1,961 @@
...
+            "type": "RoomAbridged",
+            "name": "TinyHouseZone",
+            "name_original": "Tiny House Zone",
+            "properties": {
+                "type": "RoomProperties",
+                "energy": {
+                    "type": "RoomEnergyPropertiesAbridged",
+                    "program_type": null,
+                    "construction_set": null,
+                    "people": null,
```
Comment (tanushree04) [inline tests/fixtures/1_single_zone_tiny_house.json]: @chriswmackey  would this be the name of the internal load as well?

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 11]
```diff
@@ -0,0 +1,961 @@
...
+                    "lighting": null,
+                    "electric_equipment": null,
+                    "gas_equipment": null,
+                    "infiltration": null,
+                    "ventilation": null
+                }
+            },
+            "faces": [
```
Comment (tanushree04) [inline tests/fixtures/1_single_zone_tiny_house.json]: @chriswmackey is there a need for a validation in the schema to ensure the number of faces are 6?

File: tests/fixtures/1_single_zone_tiny_house.json [hunk 12]
```diff
@@ -0,0 +1,961 @@
...
+                }
+            },
+            "face_type": {
+                "type": "Floor"
+            },
+            "boundary_condition": {
+                "type": "Ground",
+                "sun_exposure": "NoSun",
+                "wind_exposure": "NoWind",
+                "view_factor": "autocalculate"
```
Comment (tanushree04) [inline tests/fixtures/1_single_zone_tiny_house.json]: Is `autocalculate` the default value?

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 11 above)
Comment (macumber) [reply tests/fixtures/1_single_zone_tiny_house.json]: I assume there can be more than 6 faces.  Minimum number of faces for an enclosed zone would be 4, might be best to have no limits to allow unenclosed zones.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 1 above)
Comment (mostaphaRoudsari) [reply tests/fixtures/1_single_zone_tiny_house.json]: It can go either way. I think the current design protects the global / default construction from being overwritten or mixed with user / model construction sets.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 4 above)
Comment (mostaphaRoudsari) [reply tests/fixtures/1_single_zone_tiny_house.json]: yes and yes.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 3 above)
Comment (mostaphaRoudsari) [reply tests/fixtures/1_single_zone_tiny_house.json]: I also like `display_name` or `ui_name`.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 11 above)
Comment (mostaphaRoudsari) [reply tests/fixtures/1_single_zone_tiny_house.json]: > might be best to have no limits to allow unenclosed zones

@macumber curios to know what are the cases that you think an unenclosed zone will be useful? We are currently enforcing closed zones in <s>the UI</s> Honeybee.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 6 above)
Comment (mostaphaRoudsari) [reply tests/fixtures/1_single_zone_tiny_house.json]: > it would be good for us if we can differentiate between light shelves and interior partitions.

Agreed!

> Is it possible to associate light shelves with a particular window?

I agree that it should be. As @chriswmackey pointed-out it is something that we are thinking to implement. Do you already have a property for this in OpenStudio model.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 5 above)
Comment (mostaphaRoudsari) [reply tests/fixtures/1_single_zone_tiny_house.json]: In other words, to be able to make the distinction between indoor and outdoor faces and windows we need to have access to boundary condition in higher level.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 12 above)
Comment (mostaphaRoudsari) [reply tests/fixtures/1_single_zone_tiny_house.json]: Yes and if I remember correctly it is protected by default not to be overwritten. Do you think that it should not be the default value?

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 11 above)
Comment (macumber) [reply tests/fixtures/1_single_zone_tiny_house.json]: Some applications (e.g. CBECC-com) use pinwheel models for some applications.  You can't use all the shadowing methods and you have to manually input zone volumes.  I think it's just an application preference, if your users don't need it then I think requiring enclosed zones is good.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 6 above)
Comment (macumber) [reply tests/fixtures/1_single_zone_tiny_house.json]: We have an object `OS:DaylightingDevice:Shelf` that points to a `SubSurface`, `InteriorPartitionSurface` on the inside, and `ShadingSurface` on the outside.  It's similar to the `DaylightingDevice:Shelf` in E+.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 10 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: Yes, it will @tanushree04 .  Because the schema of this object is Abridged, all of the keys here should have names of other objects on them.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 11 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: @macumber ,
The Python objects for Room geometry have checks for solidity and, because of this, we intend to enforce closed volumes for all of the models created from the plugins. However, I would agree that we don't need to be strict in enforcing closed volumes within the schema validation.  People may have a model without closed volumes that they intend to edit into one that has closed volumes (I'm thinking about the case where Theo makes an editor for models in the honeybee JSON schema).  Also, if a user of Pollination API wants us to run a check for solidity, we can set up a service that re-serializes the JSON to honeybee Python objects and runs the solidity check.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 1 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: I implemented this as a named referenced in the updates samples.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 2 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: I am going to resolve this now that it has been answered.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 3 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: I have changed this to `display_name` in the updated examples.  Glad that we all agree this makes sense.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 7 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: I am just making a note of @macumber 's suggestion here if we go for the 2-way reference:

> my input on parents referencing children and children referencing parents, this is a fine convenience to serialize to JSON, I would just write a comment on one of the fields in the schema to say "this one is used in deserialization, the other one is not",  that way it is clear which one of the links actually describes the information

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 8 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: @tanushree04 , I just revised this since the keys are all optional (the samples now don't have them included).  The default values come from EnergyPlus and are as follows:
`diffuse_reflectance` = 0.2
`specular_reflectance` = 0
`transmittance` = 0
`transmittance_schedule` = None

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 9 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: Yes, @tanushree04 .  They have the exact same name restriction as the construction and material names.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 11 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: I am going to resolve this now that we agreed that keeping the schema felxible is a good strategy.

File: tests/fixtures/1_single_zone_tiny_house.json (hunk 12 above)
Comment (chriswmackey) [reply tests/fixtures/1_single_zone_tiny_house.json]: Protected in the OpenStudio SDK, @mostaphaRoudsari ?

File: app/models/samples/json/model_shoe_box.json [hunk 13]
```diff
@@ -0,0 +1,614 @@
...
+                ]
+            },
+            "properties": {
+                "type": "FaceProperties",
+                "energy": {
+                    "type": "FaceEnergyPropertiesAbridged"
+                }
+            },
+            "face_type": {
+                "type": "Floor"
```
Comment (tanushree04) [inline app/models/samples/json/model_shoe_box.json:273]: Is face_type a dictionary? ` "face_type": "Floor",`

Comment (chriswmackey) [reply app/models/samples/json/model_shoe_box.json:273]: It is implemented as a dictionary and not a string because, for some face types (like AirWalls), we may want to store some other properties within the face_type instance.  For example, AirWalls might have additional parameters like the amount of air mixing or whether the air wall should be removed for daylight simulation.  For now, do not worry about this case, @tanushree04 , and just implement them as dictionaries with a single `type` parameter.  We can discuss more in our next meeting.

Comment (tanushree04) [reply app/models/samples/json/model_shoe_box.json]: Okay. So each face_type would have its own schema?

File: app/models/samples/json/model_shoe_box.json [hunk 14]
```diff
@@ -0,0 +1,614 @@
...
+                ]
+            },
+            "properties": {
+                "type": "ApertureProperties",
+                "energy": {
+                    "type": "ApertureEnergyPropertiesAbridged"
+                }
+            },
+            "aperture_type": {
+                "type": "Window"
```
Comment (tanushree04) [inline app/models/samples/json/model_shoe_box.json:556]: Are the aperture_types:
`Window`
`OperableWindow`
`GlassDoor`
`Skylight`

Comment (chriswmackey) [reply app/models/samples/json/model_shoe_box.json:556]: `Skylight` is not an aperture type since you will know if an Aperture is a skylight if its parent face is a `RoofCeiling` or `Floor`.  The other three are all aperture types and, yes, those are the only 3.
